src/libmeasurement_kit/ooni/bouncer_impl.hpp

```diff
@@ -0,0 +1,124 @@
+
+    json request;
+    json net_tests;
+    net_tests["input-hashes"] = {};
+    net_tests["name"] = test_name;
+    net_tests["test-helpers"] = helpers;
+    net_tests["version"] = test_version;
+    request["net-tests"][0] = net_tests;
+
+    settings["http/url"] = base_bouncer_url;
+    settings["http/method"] = "POST";
+
+    http_request(
+        settings, {}, request.dump(),
+        [=](Error e, Var<http::Response> resp) {
+            if (e) {
+                reactor->call_soon([=]() { cb(e, nullptr); });
+            }
+            Var<Reply> reply(new Reply);
+            try {
+                json response = json::parse(resp->body);
+                if (response.find("error") != response.end()) {
+                    if (response["error"] == "collector-not-found") {
+                        logger->warn(
+                            "collector not found for the requested test");
+                        reactor->call_soon([=]() {
+                            cb(BouncerCollectorNotFoundError(), nullptr);
+                        });
+                        return;
+                    } else if (response["error"] == "invalid-request") {
+                        logger->warn("invalid request sent to the bouncer");
+                        reactor->call_soon([=]() {
+                            cb(BouncerInvalidRequestError(), nullptr);
+                        });
+                        return;
+                    } else {
+                        // I assume that if we receive a json with the key
+                        // "error"
+                        // then the json is not valid and we should not continue
+                        // to parse it
+                        logger->warn("bouncer generic error");
+                        reactor->call_soon(
+                            [=]() { cb(BouncerGenericError(), nullptr); });
+                        return;
+                    }
+                }
+
+                auto collector_alternate =
+                    response["net-tests"][0]["collector-alternate"];
+                bool collector_found = false;
+                for (auto collector : collector_alternate) {
+                    if (collector["type"] == "https") {
+                        reply->https_collector = collector["address"];
+                        collector_found = true;
+                        break;
+                    }
+                }
+                if (!collector_found) {
+                    logger->warn(
+                        "no https collector found in bouncer response");
+                    reactor->call_soon(
+                        [=]() { cb(MissingHttpsCollectorError(), nullptr); });
+                }
+                auto test_helpers_alternate =
+                    response["net-tests"][0]["test-helpers-alternate"]
+                            [test_name];
+                bool helper_found = false;
+                for (auto helper : test_helpers_alternate) {
```

Comment:
For example try requesting as a test_helper: `tcp-echo` (the test helper used in http-invalid-request-line).

What you will get back is not a .onion address and has no test-helpers-alternate.

Example:

```
curl --data '{"net-tests": [{"test-helpers": ["tcp-echo"], "name": "http_invalid_request_line", "input-hashes": [], "version": "1.0.0"}]}' https://bouncer.ooni.io/bouncer/net-tests
```

Response:
```
{"net-tests": [{"collector-alternate": [{"type": "https", "address": "https://a.collector.ooni.io:4441"}, {"front": "a0.awsstatic.com", "type": "cloudfront", "address": "https://das0y2z2ribx3.cloudfront.net"}], "version": "1.0.0", "name": "http_invalid_request_line", "test-helpers": {"tcp-echo": "213.138.109.232"}, "test-helpers-alternate": {}, "collector": "httpo://ihiderha53f36lsd.onion", "input-hashes": []}]}
```